src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs

```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
+
+            if (includeCast)
+            {
+                navigationPathSegments.Add(new TypeSegment(resourceContext.StructuredType, navigationSource: null));
+            }
+
+            navigationPathSegments.Add(new NavigationPropertySegment(navigationProperty, navigationSource: null));
+
+            string link = resourceContext.InternalUrlHelper.CreateODataLink(navigationPathSegments);
+            if (link == null)
+            {
+                return null;
+            }
+
+            return new Uri(link);
+        }
+
+        private static IList<ODataPathSegment> GenerateODataPathSegmentsForContainedEntity(
+            ResourceContext resourceContext, bool includeCast)
+        {
+            IList<ODataPathSegment> odataPathSegments = new List<ODataPathSegment>();
+            Stack<ODataPathSegment> reversedOdataPathSegments = new Stack<ODataPathSegment>();
+
+            ResourceContext currentLevelResourceContext = resourceContext;
+
+            // Loop through the all ExpandedResource(s).
+            // We generate OData Path Segments for all levels.
+            while (currentLevelResourceContext != null)
+            {
+                IEdmNavigationSource navigationSource = currentLevelResourceContext.NavigationSource;
+
+                // For Contained entity sets
+                if (navigationSource is IEdmContainedEntitySet containedEntitySet)
+                {
+                    NavigationPropertySegment navigationPropertySegment = new NavigationPropertySegment(containedEntitySet.NavigationProperty, containedEntitySet.ParentNavigationSource);
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    reversedOdataPathSegments.Push(navigationPropertySegment);
+                }
+
+                // For Non-Contained entity sets
+                if(navigationSource is IEdmEntitySet entitySet)
+                {
+                    EntitySetSegment entitySetSegment = new EntitySetSegment(entitySet);
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    reversedOdataPathSegments.Push(entitySetSegment);
+                }
+
+                // For Singletons
+                if(navigationSource is IEdmSingleton singleton)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    reversedOdataPathSegments.Push(singletonSegment);
+                }
+
+                currentLevelResourceContext = currentLevelResourceContext.SerializerContext.ExpandedResource;
+
+                if (includeCast && currentLevelResourceContext != null)
```

Comment:
>includeCast [](start = 20, length = 11)

we should determine this for each step depending on whether the type that the segment we are appending is defined on matches the type of the previous segment.